src/db/sysdb_search.c

```diff
@@ -543,6 +590,212 @@ errno_t sysdb_search_ts_matches(TALLOC_CTX *mem_ctx,
+        goto done;
+    }
+
+    for (size_t i = 0; i < ts_res->count; i++) {
+        dn_filter = talloc_asprintf_append(
+                                  dn_filter,
+                                  "(%s=%s)",
+                                  SYSDB_DN,
+                                  ldb_dn_get_linearized(ts_res->msgs[i]->dn));
+        if (dn_filter == NULL) {
+            ret = ENOMEM;
+            goto done;
+        }
+    }
+
+    dn_filter = talloc_asprintf_append(dn_filter, "))");
+    if (dn_filter == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    *_dn_filter = talloc_steal(mem_ctx, dn_filter);
+    ret = EOK;
+done:
+    talloc_zfree(tmp_ctx);
+    return ret;
+}
+
+errno_t sysdb_search_with_ts_attr(TALLOC_CTX *mem_ctx,
+                                  struct sss_domain_info *domain,
+                                  struct ldb_dn *base_dn,
+                                  enum ldb_scope scope,
+                                  int optflags,
+                                  const char *filter,
+                                  const char *attrs[],
+                                  struct ldb_result **_res)
+{
+    TALLOC_CTX *tmp_ctx = NULL;
+    struct ldb_result *res;
+    errno_t ret;
+    struct ldb_message **ts_msgs = NULL;
+    struct ldb_result *ts_cache_res = NULL;
+    size_t ts_count;
+    char *dn_filter = NULL;
+
+    if (filter == NULL) {
+        return EINVAL;
+    }
+
+    tmp_ctx = talloc_new(NULL);
+    if (tmp_ctx == NULL) {
+        return ENOMEM;
+    }
+
+    /* We assume that some of the attributes are more up-to-date in
+     * timestamps db and we're supposed to search by them, so let's
+     * first search the timestamp db
+     */
+    ret = sysdb_search_ts_entry(tmp_ctx, domain->sysdb, base_dn,
+                                scope, filter, attrs,
+                                &ts_count, &ts_msgs);
+    if (ret == ENOENT) {
+        ts_count = 0;
+    } else if (ret != EOK) {
+        goto done;
+    }
+
+    if (optflags & SYSDB_SEARCH_WITH_TS_ONLY_SYSDB_FILTER) {
+        /* We only care about searching the persistent db */
+        ts_cache_res = talloc_zero(tmp_ctx, struct ldb_result);
+        if (ts_cache_res == NULL) {
+            ret = ENOMEM;
+            goto done;
+        }
+        ts_cache_res->count = ts_count;
+        ts_cache_res->msgs = talloc_steal(ts_cache_res, ts_msgs);
+    } else {
+        /* Because the timestamp database does not contain all the
+         * attributes, we need to search the persistent db for each
+         * of the entries found and merge the results
+         */
+        struct ldb_result ts_res;
+
+        memset(&ts_res, 0, sizeof(struct ldb_result));
+        ts_res.count = ts_count;
+        ts_res.msgs = ts_msgs;
+
+        /* To do that we construct the list of DNs found in the TS cache */
+        ret = sysdb_dn_filter(tmp_ctx, &ts_res, &dn_filter);
+        if (ret != EOK) {
+            goto done;
+        }
+
```

Comment:
This might become quite a heavy filter. E.g. in the refresh use-case is someone calls 'sss_cache -E' on a cache with some 100 or 1000 users the `dataExpireTimestamp<=....` will return all users. I'm not sure what the limit here is but it might be more efficient to run base searches with the DNs of the results directly?